File: frontend/src/lib/totp.ts
/**
 * TOTP helpers, wrapping `otpauth` with the app's fixed parameters
 * (SHA1 / 6 digits / 30s / ±1 window). These MUST match the backend `pyotp`
 * configuration and the architecture doc, or codes will never verify.
 */
import { TOTP as OTPAuthTOTP, Secret } from 'otpauth';
import { TOTP } from '@/config/constants';

function makeTotp(secretBase32: string): OTPAuthTOTP {
  return new OTPAuthTOTP({
    issuer: 'ParadoxConnect',
    algorithm: TOTP.algorithm,
    digits: TOTP.digits,
    period: TOTP.period,
    secret: Secret.fromBase32(secretBase32),
  });
}

/** Generate the current 6-digit code for a secret (used on the student device). */
export function generateCode(secretBase32: string, at: number = Date.now()): string {
  return makeTotp(secretBase32).generate({ timestamp: at });
}

/**
 * Verify a submitted code against a secret with ±window tolerance.
 * Returns the matched time-step delta (0, ±1) or null if invalid — the delta
 * lets a caller build replay protection keyed on the exact window.
 */
export function verifyCode(
  secretBase32: string,
  code: string,
  at: number = Date.now(),
): number | null {
  const delta = makeTotp(secretBase32).validate({
    token: code,
    timestamp: at,
    window: TOTP.window,
  });
  return delta; // null when no match within the window
}

/** Milliseconds remaining in the current 30s step (for the countdown UI). */
export function secondsRemaining(at: number = Date.now()): number {
  const stepMs = TOTP.period * 1000;
  return Math.ceil((stepMs - (at % stepMs)) / 1000);
}

File: backend/app/qr/totp.py
"""TOTP helpers for the digital ID.

Parameters MUST match the frontend (`otpauth`) and `docs/api-contract.md`:
SHA1, 6 digits, 30s period, ±1 step window, Base32 160-bit secret.
"""

import hashlib
import hmac
import time

import pyotp


PERIOD_SECONDS = 30
DIGITS = 6
WINDOW = 1
REPLAY_TTL_SECONDS = PERIOD_SECONDS * (2 * WINDOW + 1)


def generate_secret() -> str:
    """A fresh Base32-encoded 160-bit secret."""
    return pyotp.random_base32()


def _totp(secret_base32: str) -> pyotp.TOTP:
    return pyotp.TOTP(
        secret_base32, digits=DIGITS, digest=hashlib.sha1, interval=PERIOD_SECONDS
    )


def verify_and_step(
    secret_base32: str, code: str, at: float | None = None
) -> int | None:
    """Verify a code within ±WINDOW steps.

    Returns the matched time-step (for replay protection) or None if invalid.
    """
    now = time.time() if at is None else at
    totp = _totp(secret_base32)
    for offset in (0, -1, 1):
        moment = now + offset * PERIOD_SECONDS
        if hmac.compare_digest(totp.at(moment), code):
            return int(moment // PERIOD_SECONDS)
    return None
